src/Datadog.Trace.ClrProfiler.Managed/CallTarget/Handlers/IntegrationMapper.cs

```diff
@@ -0,0 +1,714 @@
+                ilWriter.EmitCall(OpCodes.Call, unwrapReturnValue, null);
+            }
+
+            ilWriter.Emit(OpCodes.Ret);
+
+            Log.Debug($"Created EndMethod Dynamic Method for '{integrationType.FullName}' integration. [Target={targetType.FullName}, ReturnType={returnType.FullName}]");
+            return callMethod;
+        }
+
+        internal static DynamicMethod CreateAsyncEndMethodDelegate(Type integrationType, Type targetType, Type returnType)
+        {
+            Log.Debug($"Creating AsyncEndMethod Dynamic Method for '{integrationType.FullName}' integration. [Target={targetType.FullName}, ReturnType={returnType.FullName}]");
+            MethodInfo onAsyncMethodEndMethodInfo = integrationType.GetMethod(EndAsyncMethodName, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Static);
+            if (onAsyncMethodEndMethodInfo is null)
+            {
+                Log.Warning($"Couldn't find the method: {EndAsyncMethodName} in type: {integrationType.FullName}");
+                return null;
+            }
+
+            if (!onAsyncMethodEndMethodInfo.ReturnType.IsGenericParameter && onAsyncMethodEndMethodInfo.ReturnType != returnType)
+            {
+                throw new ArgumentException($"The return type of the method: {EndAsyncMethodName} in type: {integrationType.FullName} is not {returnType}");
+            }
+
+            Type[] genericArgumentsTypes = onAsyncMethodEndMethodInfo.GetGenericArguments();
+            if (genericArgumentsTypes.Length < 1 || genericArgumentsTypes.Length > 2)
+            {
+                throw new ArgumentException($"The method: {EndAsyncMethodName} in type: {integrationType.FullName} must have the generic type for the instance type.");
+            }
+
+            ParameterInfo[] onAsyncMethodEndParameters = onAsyncMethodEndMethodInfo.GetParameters();
+            if (onAsyncMethodEndParameters.Length < 3)
+            {
+                throw new ArgumentException($"The method: {EndAsyncMethodName} with {onAsyncMethodEndParameters.Length} paremeters in type: {integrationType.FullName} has less parameters than required.");
+            }
+            else if (onAsyncMethodEndParameters.Length > 4)
+            {
+                throw new ArgumentException($"The method: {EndAsyncMethodName} with {onAsyncMethodEndParameters.Length} paremeters in type: {integrationType.FullName} has more parameters than required.");
+            }
+
+            if (onAsyncMethodEndParameters[1].ParameterType != typeof(Exception) && onAsyncMethodEndParameters[2].ParameterType != typeof(Exception))
+            {
+                throw new ArgumentException($"The Exception type parameter of the method: {EndAsyncMethodName} in type: {integrationType.FullName} is missing.");
+            }
+
+            if (onAsyncMethodEndParameters[2].ParameterType != typeof(CallTargetState) && onAsyncMethodEndParameters[3].ParameterType != typeof(CallTargetState))
```

Comment:
Last out of bounds exception when accessing `onAsyncMethodEndParameters[3]`

Reply:
fixed in https://github.com/DataDog/dd-trace-dotnet/pull/1007/commits/04ae61a340872f36f5e93929e69b26305bc991dd